The original scripts were written to run with Debian Squeeze. File paths
for several system functions are different in Gentoo and possibly with newer
Debian systems as well.

This patch also sets different locations for the Bubba web UI installed
components.

Patch by Gordon Bos

--- a/bubba-aptd.tac
+++ b/bubba-aptd.tac
@@ -414,13 +414,13 @@
 
     # Current LAN interface
     LAN = subprocess.Popen(
-        ['bubba-networkmanager-cli', 'getlanif'],
+        ['/opt/bubba/bin/bubba-networkmanager-cli', 'getlanif'],
         stdout=subprocess.PIPE
     ).communicate()[0].strip()
 
     # current WAN interface
     WAN = subprocess.Popen(
-        ['bubba-networkmanager-cli', 'getwanif'],
+        ['/opt/bubba/bin/bubba-networkmanager-cli', 'getwanif'],
         stdout=subprocess.PIPE
     ).communicate()[0].strip()
 
@@ -497,7 +497,7 @@
     # Home partition mode
     disks = json.loads(
         subprocess.Popen(
-            ['diskmanager', 'disk', 'list'],
+            ['/opt/bubba/sbin/diskmanager', 'disk', 'list'],
             stdout=subprocess.PIPE
         ).communicate()[0].strip()
     )
--- a/bubba-easyfind.hook
+++ b/bubba-easyfind.hook
@@ -9,5 +9,5 @@
 else
         # Update easyfind db
         echo "Launching easyfind update script"
-        /usr/lib/web-admin/easyfind.pl
+        /opt/bubba/bin/easyfind.pl
 fi
--- a/bubba-easyfind.tac
+++ b/bubba-easyfind.tac
@@ -88,7 +88,7 @@
 """
 
 application = service.Application("Easyfind Update Service")
-config = ConfigObj('/etc/network/easyfind.conf', configspec=spec.split("\n"))
+config = ConfigObj('/etc/bubba/easyfind.conf', configspec=spec.split("\n"))
 validator = Validator()
 config.validate(validator, copy=True)
 try:
@@ -139,7 +139,7 @@
 
     # current WAN interface
     WAN = subprocess.Popen(
-        ['bubba-networkmanager-cli', 'getwanif'],
+        ['/opt/bubba/bin/bubba-networkmanager-cli', 'getwanif'],
         stdout=subprocess.PIPE
     ).communicate()[0].strip()
 
--- a/Makefile.PL
+++ b/Makefile.PL
@@ -2,5 +2,5 @@
 use ExtUtils::MakeMaker;
 
 WriteMakefile( DIR => [ 'web-admin' ] ,	INSTALLDIRS => 'vendor',
-	INSTALLVENDORSCRIPT => '$(PREFIX)/lib/web-admin',
-	INSTALLSITESCRIPT => '$(PREFIX)/lib/web-admin');
+	INSTALLVENDORSCRIPT => '/opt/bubba/bin',
+	INSTALLSITESCRIPT => '/opt/bubba/bin');
--- a/personal-setting-files.txt
+++ b/personal-setting-files.txt
@@ -1,22 +1,23 @@
 # network related
-/etc/network/interfaces
-/etc/network/firewall.conf
-/etc/network/easyfind.conf
+/etc/conf.d/net
+/etc/bubba/firewall.conf
+/etc/bubba/easyfind.conf
 /etc/dnsmasq.d/bubba.conf
-/etc/hostname
+/etc/conf.d/hostname
 /etc/hosts
 /etc/proftpd/proftpd.conf
 /etc/samba/smb.conf
 /etc/postfix/main.cf
-/etc/dhcp/dhclient.conf
+/etc/postfix/master.cf
+/etc/dhcpcd.conf
 /etc/cups/printers.conf
 /etc/resolv.conf
 /etc/fetchmailrc
 /etc/postfix/sasl_passwd
 /etc/postfix/sasl_passwd.db
-/etc/ftd/ftdconfig.ini
+/etc/bubba/ftdconfig.ini
 /etc/hostapd/hostapd.conf
-/etc/bubba-networkmanager.conf
+/etc/bubba/networkmanager.conf
 
 /etc/minidlna.conf
 
@@ -41,8 +42,8 @@
 /etc/apache2/cacert.pem
 /etc/apache2/privkey.pem
 
-# Squeezecenter
-/var/lib/squeezeboxserver/prefs
+# LogitechMediaServer
+/etc/squeezeboxserver
 
 # Bubba diverted configs
 /etc/bubba/*
--- a/web-admin/bin/adminfunctions.php
+++ b/web-admin/bin/adminfunctions.php
@@ -2,7 +2,7 @@
 # Setting the ctype locale to en_US.UTF-8, mostly to enamble escapeshellargs to function properly
 setlocale( LC_CTYPE, 'en_US.UTF-8' );
 define("DEBUG",0);
-define("BUBBA_EASYFIND_CONF","/etc/network/easyfind.conf");
+define("BUBBA_EASYFIND_CONF","/etc/bubba/easyfind.conf");
 class AdminException extends Exception {
 
 	const MYSQL_CONNECT_ERROR = 0x01;
@@ -16,7 +16,7 @@
 function _getlanif(){
 	static $lanif="";
 	if($lanif==""){
-		$lanif=shell_exec("/usr/bin/bubba-networkmanager-cli getlanif");
+		$lanif=shell_exec("/opt/bubba/bin/bubba-networkmanager-cli getlanif");
 	}
 	return rtrim($lanif);
 }
@@ -430,12 +430,12 @@
 
 function get_interface_info($iface){
 	$res=array();
-	exec("/sbin/ifconfig $iface",$out,$ret);
+	exec("/bin/ifconfig $iface",$out,$ret);
 	foreach($out as $line){
-		if(preg_match("/inet addr:([\d.]+)/",$line,$match)){
+		if(preg_match("/inet ([\d.]+)/",$line,$match)){
 			$res[0]=$match[1];
 		}
-		if(preg_match("/Mask:([\d.]+)/",$line,$match)){
+		if(preg_match("/netmask ([\d.]+)/",$line,$match)){
 			$res[1]=$match[1];
 		}
 	}
@@ -467,17 +467,19 @@
 	return $res;
 }
 
+/*
+	// Gordon: 2015-07-15 - does not appear to be used any more
 function _check_dhcp($iface=""){
 	if($iface=""){
 		$iface=_getlanif();
 	}
 	$cdhcp=false;
-	$netcfg=file("/etc/network/interfaces");
+	$netcfg=file("/etc/conf.d/net");
 	foreach ($netcfg as $i) {
 		$trim_line = trim($i);
-		$pieces = explode(" ",$trim_line);
+		$pieces = explode("=",$trim_line);
 		if(count($pieces)==4){
-			if($pieces[1]==$iface && $pieces[3]=="dhcp"){
+			if($pieces[1]=="config_$iface" && $pieces[3]=="\"dhcp\""){
 				$cdhcp=true;
 				break;
 			}
@@ -513,7 +515,7 @@
 			}			
 		}
 	} else { // default
-		exec("/sbin/ifconfig $iface",$out,$ret);
+		exec("/bin/ifconfig $iface",$out,$ret);
 		foreach($out as $line){
 			if(preg_match("/inet addr:([\d.]+)/",$line,$match)){
 				$res[0]=$match[1];
@@ -523,7 +525,7 @@
 			}
 		}
 	}
-	exec("/sbin/route -n",$out,$ret);
+	exec("/bin/route -n",$out,$ret);
 	$res[2]="0.0.0.0";
 	foreach($out as $line){
 		if(preg_match("/^0\.0\.0\.0\s+([\d.]+)/",$line,$match)){
@@ -543,6 +545,7 @@
 	return $res;	
 	
 }
+*/
 
 function set_static_netcfg($iface, $ip,$nm,$gw){
    
@@ -615,7 +618,7 @@
 
 function query_service($name){
 
-   $res=glob("/etc/rc2.d/S??$name");
+   $res=glob("/etc/runlevels/default/$name");
    return $res?true:false;
 
 }
@@ -1231,8 +1234,8 @@
 	exec($cmd,$out,$ret);
 	$versions = array();
 	foreach( $out as $line ) {
-		list( $name, $version ) = explode( ' ', $line );
-		$versions[$name] = $version;
+		list( $name, $version ) = explode( ' ', $line." null" );
+		$versions["$name"] = $version;
 	}
 
 	if( count($versions) == 1 ) {
@@ -1299,7 +1302,7 @@
 
 function get_timezone_info() {
 	
-	$zoneinfo = '/usr/share/zoneinfo/right';
+	$zoneinfo = '/usr/share/zoneinfo';
 	if ($h_zonebase = opendir($zoneinfo)) {
 		
 		$zones = array();
--- a/web-admin/bin/backup.pl
+++ b/web-admin/bin/backup.pl
@@ -30,7 +30,7 @@
 use constant NCFTPLS          => "/usr/bin/ncftpls";
 use constant MSGFILE          => "/tmp/backup.README";
 use constant SSH              => "/usr/bin/ssh";
-use constant SCRIPTNAME       => "/usr/lib/web-admin/backup.pl";
+use constant SCRIPTNAME       => "/opt/bubba/bin/backup.pl";
 use constant INCLUDEFILE      => "includeglob.list";
 use constant INCLUDE_CHILD    => "include_childglob.list";
 use constant EXCLUDEFILE      => "excludeglob.list";
@@ -38,10 +38,10 @@
 use constant CHECK_CHILD      => 1;
 use constant CHECK_PARENT     => 0;
 use constant CRON_FILE        => "/etc/cron.d/bubba-backup";
-use constant QUEUE_FILE       => "/etc/bubba-backup.queue";
+use constant QUEUE_FILE       => "/var/lib/bubba/bubba-backup.queue";
 use constant LOCK_FILE        => "/var/lock/backup.lock";
 use constant LOCK_RESTOREFILE => "/var/lock/restore.lock";
-use constant DISKMANAGER      => "/usr/sbin/diskmanager";
+use constant DISKMANAGER      => "/opt/bubba/sbin/diskmanager";
 
 use constant DEBUG => 0;
 
--- a/web-admin/bin/diskdaemon.pl
+++ b/web-admin/bin/diskdaemon.pl
@@ -25,9 +25,9 @@
 use POSIX ":sys_wait_h";
 use IPC::SysV qw(IPC_CREAT IPC_RMID S_IRUSR ftok);
 
-use constant SOCKNAME		=> "/tmp/bubba-disk.socket";
-use constant PIDFILE		=> '/tmp/bubba-disk.pid';
-use constant MANAGER		=> '/usr/sbin/diskmanager';
+use constant SOCKNAME		=> "/run/bubba-disk.socket";
+use constant PIDFILE		=> '/run/bubba-disk.pid';
+use constant MANAGER		=> '/opt/bubba/sbin/diskmanager';
 use constant DELAY          => 20;
 
 my $daemon = Proc::Daemon->new(
--- a/web-admin/bin/easyfind.pl
+++ b/web-admin/bin/easyfind.pl
@@ -8,7 +8,7 @@
 
 use strict;
 use constant WAN_IF => "eth0";
-use constant EASYFIND_CONF => "/etc/network/easyfind.conf";
+use constant EASYFIND_CONF => "/etc/bubba/easyfind.conf";
 use constant KEY => "/etc/network/bubbakey";
 use constant BOOTARGS => "/proc/cmdline";
 
--- a/web-admin/lib/Bubba.pm
+++ b/web-admin/lib/Bubba.pm
@@ -416,13 +416,9 @@
    my ($name)=@_;
    my $ret=0;
 
-   system("userdel $name");
+   system("smbpasswd -x $name &>/dev/null && userdel $name");
    $ret=$?;
    
-   if ($ret==0) {
-      system("smbpasswd -x $name");
-   }
-
    return $ret;
 }
 
@@ -532,6 +528,28 @@
 
 }
 
+# Gordon : 2015-06-22 - added function to keep existing hosts entries
+sub update_hostsfile {
+
+	my ($lanip,$name) = @_;	
+	if ($lanip=="127.0.0.1") { return 0;}
+	my $oldname;
+	use File::Slurp;
+	my $hosts = read_file('/etc/hosts');
+	my @lines = split("\n", $hosts);
+	chomp(@lines);
+	foreach (@lines) {
+		if ( /^$lanip(.+)$/ ) {
+			/\s([^\s\.]+)[\s\.]/;
+			$oldname=$1;
+		}
+	}
+	$hosts =~ s/\$oldname/$lanip/g;
+	write_file( '/etc/hosts', $hosts );
+
+}
+
+
 # Change hostname
 #
 # Args   : name - New hostname  
@@ -544,14 +562,15 @@
 	my ($name)=@_;
 
 	system("echo $name > /proc/sys/kernel/hostname");
-	system("echo $name > /etc/hostname");
-	system("echo $name.localdomain > /etc/mailname");
+	system("/bin/sed -i \"s/\s*\(hostname=\).*$/\\1\\\"$name\\\"/\"   /etc/conf.d/hostname");
+#	system("echo $name.localdomain > /etc/mailname");
 
 	%ifs = read_interfaces();
 	chomp($lan = _get_lanif);
 	$lanip = $ifs{$lan}{"options"}{"address"};
 	$lanip = '127.0.0.1' unless $lanip;
-	write_hostsfile($lanip,$name);
+#	write_hostsfile($lanip,$name);
+	update_hostsfile($lanip,$name);
 
 	if(!query_service("dnsmasq")){
 		#restart dnsmasq
@@ -564,10 +583,10 @@
 	system("echo send host-name \\\"$name\\\"\\\; >> /etc/dhcp/dhclient.conf.new");
 	system("mv /etc/dhcp/dhclient.conf.new /etc/dhcp/dhclient.conf");
 	$lan = _get_lanif;
-	system("/sbin/ifup --force eth0 $lan");
+	system("/usr/bin/rc-config restart `rc-config list default | grep "^\s*net\."`");
 
 	if(change_ftp_servername($name)){
-		system("/etc/init.d/proftpd restart");
+		system("/usr/bin/rc-config restart proftpd");
 	}
 
 	restart_avahi();
@@ -589,7 +608,7 @@
 sub power_off{
 	use Bubba::Info;
 	if(isB3()){
-		system("/sbin/write-magic 0xdeadbeef");
+		system("/opt/bubba/bin/write-magic 0xdeadbeef");
 		return system("/sbin/reboot");
 	}elsif(isB2()){
 		if( -e "/sys/devices/platform/bubbatwo/magic" ){
@@ -639,9 +658,9 @@
    my ($if)=@_;
    my $ret;
 
-   $ret=system("/sbin/ifdown $if");
+   $ret=system("/usr/bin/rc-config stop net.$if");
    if ($ret==0) {
-      $ret=system("/sbin/ifup $if");
+      $ret=system("/usr/bin/rc-config start net.$if");
    }
    return $ret;
 }
@@ -769,7 +788,25 @@
 sub set_nameserver{
 	my ($ns)=@_;
 	
-	return system("echo -ne 'search\nnameserver $ns\n'>/etc/resolv.conf");
+#	return system("echo -ne 'search\nnameserver $ns\n'>/etc/resolv.conf");
+# Gordon : 2015-06-22 Don't delete domain information in this file
+	my $findstring="nameserver";
+	my $ret;
+        use File::Slurp;
+        my $resolvconf = read_file('/etc/resolv.conf');
+        my @lines = split("\n", resolvconf);
+        chomp(@lines);
+        foreach (@lines) {
+                if ( /^$findstring/ ) {
+                        $ret .= "nameserver $ns\n";
+			$findstring="-";
+                }else {
+			$ret .= $_"\n";
+		}
+        }
+        
+        write_file( '/etc/resolv.conf', $ret );
+
 }
 
 # Is service running?
@@ -785,13 +822,13 @@
 	my $pidfile;
 	
 	if ($service eq "fetchmail"){
-		$pidfile="/var/run/fetchmail/.fetchmail";
-	} elsif ($service eq "dnsmasq") {
-		$pidfile="/var/run/dnsmasq/dnsmasq.pid";
+		$pidfile="/var/run/fetchmail/fetchmail.pid";
 	} elsif ($service eq "avahi-daemon"){
 		$pidfile="/var/run/avahi-daemon/pid";
 	} elsif ($service eq "tor"){
 		$pidfile="/var/run/tor/tor.pid";
+	} elsif ($service eq "filetransferdaemon"){
+		$pidfile="/var/run/ftd.pid";
 	} else {
 		$pidfile = "/var/run/$service.pid";
 	}
@@ -869,7 +906,7 @@
 sub add_service{
    my ($service)=@_;
    
-   return system("/usr/sbin/update-rc.d $service defaults");  
+   return system("/sbin/rc-update add $service default");
 }
 
 # Add service att specific init "level"
@@ -881,7 +918,7 @@
 sub add_service_at_level{
    my ($service, $level)=@_;
 
-   return system("/usr/sbin/update-rc.d $service defaults $level");
+   return system("/sbin/rc-update add $service default");
 }
 
 # Remove service
@@ -892,7 +929,7 @@
 sub remove_service{
    my ($service)=@_;
    
-   return system("/usr/sbin/update-rc.d -f $service remove");  
+   return system("/sbin/rc-update del $service default");
 }
 
 # Query service
@@ -903,7 +940,7 @@
 sub query_service{
    my ($service)=@_;
 
-   return system("ls /etc/rc2.d/S??$service 1>/dev/null 2>/dev/null");
+   return system("/bin/ls /etc/runlevels/default/$service 1>/dev/null 2>/dev/null");
 
 }
 
@@ -1064,6 +1101,11 @@
 			print FILE "\tuser '".@{$line}[2]."' there with password '".@{$line}[3]."' is '".@{$line}[4]."' here ".@{$line}[5]." $keep\n";
    }
    close FILE;
+   # Gordon 2015-12-06: set correct owner and rights on fetchmailrc
+   chmod 0600, "/etc/fetchmailrc";
+   my $login, $passwd, $uid, $gid;
+   ($login,$pass,$uid,$gid) = getpwnam("fetchmail") or die "fetchmail not in passwd file";
+   chown $uid, $gid, "/etc/fetchmailrc";
 }
 
 
@@ -1366,7 +1408,7 @@
 
 	my $tempdir = tempdir( CLEANUP => 1 );
 
-	my $psettings = "/usr/share/bubba-configs/personal-setting-files.txt";
+	my $psettings = "/var/lib/bubba/personal-setting-files.txt";
 	chomp(my @psettings_data = read_file( $psettings ));
 	my $timestring = strftime "%Y-%m-%dT%H%M%S", gmtime;
 	my $filename = "$path/bubbatwo-$timestring.backup";
@@ -1401,8 +1443,8 @@
 
 	# services, boolean such if service enabled or not
 	my %services = map {
-		$_ => (defined bsd_glob "/etc/rc2.d/S??$_");
-	} qw(proftpd forked-daapd ntp filetransferdaemon cups postfix dovecot fetchmail minidlna dnsmasq logitechmediaserver hostapd netatalk ifup-br0 samba);
+		$_ => (defined -e "/etc/runlevels/default/$_");
+	} qw(proftpd forked-daapd ntpd filetransferdaemon cupsd postfix dovecot fetchmail minidlna dnsmasq logitechmediaserver hostapd netatalk net.br0 samba);
 
 	my $meta = {
 		version => $revision,
@@ -1618,8 +1660,8 @@
 			'--absolute-name'
 		);
 
-		system("/sbin/iptables-restore","/etc/network/firewall.conf");
-		system("/bin/cp","/etc/hostname","/proc/sys/kernel/hostname");
+		system("/sbin/iptables-restore","/etc/bubba/firewall.conf");
+		system("/usr/bin/rc-config","restart","hostname");
 
 		restart_network("eth0");
 		# hostapd needs to be started prior to restarting LANIF
@@ -1741,9 +1783,9 @@
 			$lan = _get_lanif;
 			restart_network($lan);
 
-			system("/sbin/iptables-restore < /etc/network/firewall.conf");
+			system("/sbin/iptables-restore < /etc/bubba/firewall.conf");
 
-			system("echo `cat /etc/hostname` > /proc/sys/kernel/hostname");
+			system("/usr/bin/rc-config","restart","hostname");
 
 			if($lines=~/proftpd/){
 				start_service("proftpd");
@@ -1795,12 +1837,12 @@
 				remove_service("fetchmail");
 			}
 
-			if($lines=~/cups/){
-				start_service("cups");
-				reload_service("cups");
+			if($lines=~/cupsd/){
+				start_service("cupsd");
+				reload_service("cupsd");
 			}else{
-				stop_service("cups");
-				remove_service("cups");
+				stop_service("cupsd");
+				remove_service("cupsd");
 			}
 
 			if($lines=~/dnsmasq/){
@@ -1872,7 +1914,7 @@
 	if (!$cmd) {
 		$cmd = "";
 	}
-	return system("/usr/lib/web-admin/easyfind.pl $cmd $name");
+	return system("/opt/bubba/bin/easyfind.pl $cmd $name");
 	
 }
 
@@ -1999,7 +2041,7 @@
 }
 
 sub _get_lanif {
-	return `/usr/bin/bubba-networkmanager-cli getlanif`;
+	return `/opt/bubba/bin/bubba-networkmanager-cli getlanif`;
 }
 
 sub _notify_read_config {
--- a/web-admin/Makefile.PL
+++ b/web-admin/Makefile.PL
@@ -8,7 +8,6 @@
 	'bin/backend.pl',
 	'bin/disk.pl',
 	'bin/firewall.pl',
-	'bin/easyfind.pl',
 	'bin/backup.pl',
 	'bin/diskdaemon.pl',
 	'bin/notify-dispatcher.pl',
--- a/web-admin/MANIFEST
+++ b/web-admin/MANIFEST
@@ -2,14 +2,11 @@
 MANIFEST
 lib/Bubba.pm
 bin/adminfunctions.php
-bin/updatebackend.pl
-bin/adminfunctions.php
 bin/disk.pl
 bin/firewall.pl
 bin/updater.pl
 bin/thumbnail.pl
 bin/backend.pl
-bin/print.pl
 bin/backup.pl
 bin/diskdaemon.pl
 bin/notify-dispatcher.pl
